src/flb_engine.c

```diff
@@ -522,6 +525,34 @@ int flb_engine_start(struct flb_config *config)
     }
     config->evl = evl;
 
+    /* Create the bucket queue (FLB_ENGINE_PRIORITY_COUNT priorities) */
+    evl_bktq = flb_bucket_queue_create(FLB_ENGINE_PRIORITY_COUNT);
+    if (!evl_bktq) {
+        return -1;
+    }
+    config->evl_bktq = evl_bktq;
+
+    /*
+     * Event loop channel to ingest flush events from flb_engine_flush()
+     *
+     *  - FLB engine uses 'ch_self_events[1]' to dispatch tasks to self
+     *  - Self to receive message on ch_parent_events[0]
+     *
+     * The mk_event_channel_create() will attach the pipe read end ch_self_events[0]
+     * to the local event loop 'evl'.
+     */
+    ret = mk_event_channel_create(config->evl,
+                                    &config->ch_self_events[0],
+                                    &config->ch_self_events[1],
+                                    &config->event_thread_init);
```

Comment:
This is a new system to deal with coroutines on the main event loop. Previously, coroutines directly handled in flush. This may cause blockage of dispatching tasks to threads if say: flush([[engine coro ~ 1s][worker coro ~ 1s]]) Previously flush would process engine_coro in place, and then after 1 second dispatch coro to worker. Now flush will dispatch both events. engine coro to self through this channel and worker coro to worker through an existing worker channel. Dispatch/flush takes highest priority.

Retries do not block dispatch either since retries are queued with schedule_request_now a timer event of default priority gives preference to flush/dispatch.